kv.go

```diff
@@ -0,0 +1,387 @@
+		Description:       cfg.Description,
+		Subjects:          []string{fmt.Sprintf(kvSubjectsTmpl, cfg.Bucket)},
+		MaxMsgsPerSubject: history,
+		MaxMsgs:           cfg.MaxMsgs,
+		MaxBytes:          cfg.MaxBytes,
+		MaxAge:            cfg.MaxAge,
+		MaxMsgSize:        cfg.MaxValueSize,
+		Storage:           cfg.Storage,
+		Replicas:          replicas,
+	}
+
+	if _, err := jsc.AddStream(scfg); err != nil {
+		return nil, err
+	}
+
+	kv := &kvs{
+		name:   cfg.Bucket,
+		stream: scfg.Name,
+		pre:    fmt.Sprintf(kvSubjectsPreTmpl, cfg.Bucket),
+		nc:     nc,
+		js:     jsc.(*js),
+	}
+	return kv, nil
+}
+
+type kvs struct {
+	name   string
+	stream string
+	pre    string
+	nc     *Conn
+	js     *js
+}
+
+// Get returns the latest value for the key.
+func (kv *kvs) Get(key string) (value []byte, revision uint64, err error) {
+	// Build by hand since simple and avoids stdlib JSON.
+	var b strings.Builder
+	b.WriteString("{\"last_by_subj\":\"")
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+	b.WriteString("\"}")
+
+	o, cancel, err := getJSContextOpts(kv.js.opts)
+	if err != nil {
+		return nil, 0, err
+	}
+	if cancel != nil {
+		defer cancel()
+	}
+
+	// Send request.
+	subj := kv.js.apiSubj(fmt.Sprintf(apiMsgGetT, kv.stream))
+	r, err := kv.nc.RequestWithContext(o.ctx, subj, []byte(b.String()))
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// FIXME(dlc) - Be good to avoid stdlib JSON when possible.
+	// Maybe: https://github.com/goccy/go-json
+	var resp apiMsgGetResponse
+	if err := json.Unmarshal(r.Data, &resp); err != nil {
+		return nil, 0, err
+	}
+	if resp.Error != nil {
+		return nil, 0, errors.New(resp.Error.Description)
+	}
+
+	return resp.Message.Data, resp.Message.Sequence, nil
+}
+
+// Put will place the new value for the key into the store.
+func (kv *kvs) Put(key string, value []byte) (revision uint64, err error) {
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+
+	pa, err := kv.js.Publish(b.String(), value)
+	if err != nil {
+		return 0, err
+	}
+	return pa.Sequence, err
+}
+
+// Create will add the key/value pair iff it does not exist.
+func (kv *kvs) Create(key string, value []byte) (revision uint64, err error) {
+	return kv.Update(key, value, 0)
+}
+
+// Update will update the value iff the latest revision matches.
+func (kv *kvs) Update(key string, value []byte, revision uint64) (uint64, error) {
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+
+	m := Msg{Subject: b.String(), Header: Header{}, Data: value}
+	m.Header.Set(ExpectedLastSubjSeqHdr, strconv.FormatUint(revision, 10))
+
+	pa, err := kv.js.PublishMsg(&m)
+	if err != nil {
+		return 0, err
+	}
+	return pa.Sequence, err
+}
+
+// Delete the key and all revisions.
+func (kv *kvs) Delete(key string) error {
+	o, cancel, err := getJSContextOpts(kv.js.opts)
+	if err != nil {
+		return err
+	}
+	if cancel != nil {
+		defer cancel()
+	}
+
+	// Build by hand since simple and avoids stdlib JSON.
```

Comment:
simple yes, but with no checking of validity of key also trivial to break based on value of `key`